File: sdk/go/runner.go
package compliance

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// RunnerConfig contains configuration for test execution
type RunnerConfig struct {
	StopOnFirstFailure bool
	ValidatePlans      bool
	CompareResults     bool
	MaxParallelTests   int
	Timeout            time.Duration
}

// DefaultRunnerConfig returns a default configuration
func DefaultRunnerConfig() RunnerConfig {
	return RunnerConfig{
		StopOnFirstFailure: false,
		ValidatePlans:      true,
		CompareResults:     true,
		MaxParallelTests:   1, // Sequential by default
		Timeout:            5 * time.Minute,
	}
}

// ProgressCallback is called for each test execution
type ProgressCallback func(testID string, current, total int)

// ComplianceRunner executes compliance tests against an engine
type ComplianceRunner struct {
	engine   ComplianceEngine
	config   RunnerConfig
	progress ProgressCallback
}

// NewComplianceRunner creates a new runner with default config
func NewComplianceRunner(engine ComplianceEngine) *ComplianceRunner {
	return &ComplianceRunner{
		engine: engine,
		config: DefaultRunnerConfig(),
	}
}

// NewComplianceRunnerWithConfig creates a runner with custom config
func NewComplianceRunnerWithConfig(engine ComplianceEngine, config RunnerConfig) *ComplianceRunner {
	return &ComplianceRunner{
		engine: engine,
		config: config,
	}
}

// SetProgressCallback sets the progress callback function
func (r *ComplianceRunner) SetProgressCallback(callback ProgressCallback) {
	r.progress = callback
}

// RunTestSuite runs all tests in a test suite
func (r *ComplianceRunner) RunTestSuite(ctx context.Context, suite *TestSuite) (*ComplianceReport, error) {
	report := NewComplianceReport()
	total := suite.Size()

	// Initialize engine
	if err := r.engine.Initialize(ctx); err != nil {
		return nil, fmt.Errorf("failed to initialize engine: %w", err)
	}
	defer r.engine.Shutdown(ctx)

	if r.config.MaxParallelTests > 1 {
		// Parallel execution
		return r.runParallel(ctx, suite, total)
	}

	// Sequential execution
	for i, testCase := range suite.TestCases {
		if r.progress != nil {
			r.progress(testCase.ID, i+1, total)
		}

		result, err := r.RunTestCase(ctx, testCase)
		if err != nil {
			result = NewComplianceResult(testCase.ID, TestStatusError).
				WithError(err.Error())
		}

		report.AddResult(result)

		if r.config.StopOnFirstFailure && !result.IsPassed() {
			break
		}
	}

	return report, nil
}

// runParallel executes tests in parallel
func (r *ComplianceRunner) runParallel(ctx context.Context, suite *TestSuite, total int) (*ComplianceReport, error) {
	report := NewComplianceReport()
	var mu sync.Mutex
	var wg sync.WaitGroup

	// Create a semaphore to limit concurrency
	sem := make(chan struct{}, r.config.MaxParallelTests)
	current := 0

	for _, testCase := range suite.TestCases {
		wg.Add(1)
		go func(tc *TestCase) {
			defer wg.Done()

			// Acquire semaphore
			sem <- struct{}{}
			defer func() { <-sem }()

			mu.Lock()
			current++
			idx := current
			mu.Unlock()

			if r.progress != nil {
				r.progress(tc.ID, idx, total)
			}

			result, err := r.RunTestCase(ctx, tc)
			if err != nil {
				result = NewComplianceResult(tc.ID, TestStatusError).
					WithError(err.Error())
			}

			mu.Lock()
			report.AddResult(result)
			mu.Unlock()
		}(testCase)
	}

	wg.Wait()
	return report, nil
}

// RunTestCase runs a single test case
func (r *ComplianceRunner) RunTestCase(ctx context.Context, testCase *TestCase) (*ComplianceResult, error) {
	// Check if engine can run this test
	if !r.engine.CanRunTest(testCase.ID) {
		return NewComplianceResult(testCase.ID, TestStatusSkipped).
			WithError("Test skipped by engine"), nil
	}

	// Create context with timeout
	if r.config.Timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, r.config.Timeout)
		defer cancel()
	}

	// Validate plan if configured
	if r.config.ValidatePlans {
		validationResult, err := r.engine.ValidatePlan(ctx, testCase.PlanBytes)
		if err != nil {
			return nil, fmt.Errorf("validation error: %w", err)
		}
		if !validationResult.IsPassed() {
			return validationResult, nil
		}
	}

	// Execute the plan
	start := time.Now()
	result, err := r.engine.ExecutePlan(ctx, testCase.PlanBytes, testCase.InputData)
	if err != nil {
		return nil, fmt.Errorf("execution error: %w", err)
	}
	result.ExecutionTime = time.Since(start)

	// Compare results if configured and expected output exists
	if r.config.CompareResults && testCase.ExpectedOutput != nil && result.OutputData != nil {
		if !r.compareOutputs(result.OutputData, testCase.ExpectedOutput) {
			result.Status = TestStatusFailed
			result.ErrorMessage = "Output does not match expected result"
		}
	}

	return result, nil
}

// compareOutputs compares actual and expected outputs
func (r *ComplianceRunner) compareOutputs(actual, expected *TableData) bool {
	// Simple comparison - can be enhanced with more sophisticated logic
	if actual.RowCount() != expected.RowCount() {
		return false
	}
	if actual.ColumnCount() != expected.ColumnCount() {
		return false
	}

	// TODO: Implement detailed cell-by-cell comparison
	// with type-aware comparison and tolerance for floating point

	return true
}

// RunnerBuilder provides a fluent interface for creating runners
type RunnerBuilder struct {
	engine   ComplianceEngine
	config   RunnerConfig
	progress ProgressCallback
}

// NewRunnerBuilder creates a new builder
func NewRunnerBuilder(engine ComplianceEngine) *RunnerBuilder {
	return &RunnerBuilder{
		engine: engine,
		config: DefaultRunnerConfig(),
	}
}

// StopOnFailure sets whether to stop on first failure
func (b *RunnerBuilder) StopOnFailure(stop bool) *RunnerBuilder {
	b.config.StopOnFirstFailure = stop
	return b
}

// ValidatePlans sets whether to validate plans before execution
func (b *RunnerBuilder) ValidatePlans(validate bool) *RunnerBuilder {
	b.config.ValidatePlans = validate
	return b
}

// CompareResults sets whether to compare results with expected output
func (b *RunnerBuilder) CompareResults(compare bool) *RunnerBuilder {
	b.config.CompareResults = compare
	return b
}

// Parallel sets the number of parallel test executions
func (b *RunnerBuilder) Parallel(threads int) *RunnerBuilder {
	b.config.MaxParallelTests = threads
	return b
}

// Timeout sets the timeout for each test
func (b *RunnerBuilder) Timeout(timeout time.Duration) *RunnerBuilder {
	b.config.Timeout = timeout
	return b
}

// OnProgress sets the progress callback
func (b *RunnerBuilder) OnProgress(callback ProgressCallback) *RunnerBuilder {
	b.progress = callback
	return b
}

// Build creates the runner
func (b *RunnerBuilder) Build() *ComplianceRunner {
	runner := NewComplianceRunnerWithConfig(b.engine, b.config)
	if b.progress != nil {
		runner.SetProgressCallback(b.progress)
	}
	return runner
}

// Made with Bob

File: sdk/cpp/src/runner.cpp
// runner.cpp — ComplianceRunner implementation

#include "substrait_compliance/runner.h"
#include <stdexcept>

namespace substrait::compliance {

// ---------------------------------------------------------------------------
// ComplianceRunner
// ---------------------------------------------------------------------------

ComplianceReport ComplianceRunner::run_test_suite(const TestSuite& suite) {
    ComplianceReport report;
    size_t total = suite.test_cases().size();
    size_t idx = 0;

    engine_->initialize();

    for (const auto& tc : suite.test_cases()) {
        notify_progress(tc.id(), ++idx, total);
        ComplianceResult result = run_test_case(tc);
        report.add_result(std::move(result));

        if (config_.stop_on_first_failure &&
            report.results().back().is_failed()) {
            break;
        }
    }

    engine_->shutdown();
    return report;
}

ComplianceReport ComplianceRunner::run_test_suite(
        const TestSuite& suite,
        ProgressCallback callback) {
    progress_callback_ = std::move(callback);
    return run_test_suite(suite);
}

ComplianceResult ComplianceRunner::run_test_case(const TestCase& tc) {
    // Optional: let engine skip certain tests
    if (!engine_->can_run_test(tc.id())) {
        return ComplianceResult(tc.id(), TestStatus::SKIPPED)
            .with_error("Engine declined this test");
    }

    // Validate first (if configured)
    if (config_.validate_plans) {
        ComplianceResult val = validate_test_plan(tc);
        if (val.is_unsupported()) {
            return val; // UNSUPPORTED → skip
        }
    }

    // Execute with timing
    ComplianceResult exec = execute_with_timing(tc);
    if (!exec.is_passed()) {
        return exec;
    }

    // Compare against expected output (if configured and expected exists)
    if (config_.compare_results && tc.expected_output().has_value()) {
        return compare_outputs(tc, exec);
    }

    // No expected output: skip rather than pass to be honest
    if (!tc.expected_output().has_value()) {
        ComplianceResult skipped(tc.id(), TestStatus::SKIPPED);
        skipped.with_error("No expected output — cannot verify correctness");
        if (exec.output_data().has_value()) {
            skipped.with_output(*exec.output_data());
        }
        skipped.with_execution_time(exec.execution_time_ms());
        return skipped;
    }

    return exec;
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

ComplianceResult ComplianceRunner::execute_with_timing(const TestCase& tc) {
    using clock = std::chrono::steady_clock;
    auto start = clock::now();

    ComplianceResult result("", TestStatus::ERROR);
    try {
        result = engine_->execute_plan(tc.plan_bytes(), tc.input_data());
        result = ComplianceResult(tc.id(), result.status())
            .with_execution_time(
                std::chrono::duration_cast<std::chrono::milliseconds>(
                    clock::now() - start).count());
        if (result.status() == TestStatus::PASSED && result.output_data().has_value()) {
            // preserve output from the original call
        }
        // Re-run to get output attached properly
        result = engine_->execute_plan(tc.plan_bytes(), tc.input_data());
    } catch (const std::exception& e) {
        auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
            clock::now() - start).count();
        return ComplianceResult(tc.id(), TestStatus::ERROR)
            .with_error(e.what())
            .with_execution_time(elapsed);
    }

    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        clock::now() - start).count();

    // Build a clean result tagged with this test's id
    ComplianceResult r(tc.id(), result.status());
    r.with_execution_time(elapsed);
    if (result.output_data().has_value()) {
        r.with_output(*result.output_data());
    }
    if (result.error_message().has_value()) {
        r.with_error(*result.error_message());
    }
    return r;
}

ComplianceResult ComplianceRunner::validate_test_plan(const TestCase& tc) {
    try {
        ComplianceResult val = engine_->validate_plan(tc.plan_bytes());
        if (val.status() == TestStatus::UNSUPPORTED) {
            return ComplianceResult(tc.id(), TestStatus::UNSUPPORTED)
                .with_error(val.error_message().value_or("Plan not supported"));
        }
        return ComplianceResult(tc.id(), TestStatus::PASSED);
    } catch (const std::exception& e) {
        // Treat a thrown exception from validate as unsupported
        return ComplianceResult(tc.id(), TestStatus::UNSUPPORTED)
            .with_error(e.what());
    }
}

ComplianceResult ComplianceRunner::compare_outputs(
        const TestCase& tc,
        const ComplianceResult& exec_result) {
    const TableData& expected = *tc.expected_output();
    const TableData& actual = exec_result.output_data().has_value()
        ? *exec_result.output_data()
        : TableData{};

    ComparisonResult cmp = comparator_->compare_tables(actual, expected);

    if (cmp.matches) {
        ComplianceResult r(tc.id(), TestStatus::PASSED);
        r.with_execution_time(exec_result.execution_time_ms());
        r.with_output(actual);
        return r;
    }

    ComplianceResult r(tc.id(), TestStatus::FAILED);
    r.with_error(cmp.message);
    r.with_execution_time(exec_result.execution_time_ms());
    if (exec_result.output_data().has_value()) {
        r.with_output(*exec_result.output_data());
    }
    return r;
}

void ComplianceRunner::notify_progress(
        const std::string& test_id,
        size_t current,
        size_t total) {
    if (progress_callback_) {
        progress_callback_(test_id, current, total);
    }
}

} // namespace substrait::compliance
